samples/bluetooth/nrf_dm/src/peer.c

```diff
@@ -0,0 +1,398 @@
+	printk("\tQuality: %s\n", quality[result->quality]);
+
+	printk("\tDistance estimates: ");
+	if (result->ranging_mode == DM_RANGING_MODE_RTT) {
+		printk("rtt: rtt=%.2f\n", result->dist_estimates.rtt.rtt);
+	} else {
+		printk("mcpd: ifft=%.2f phase_slope=%.2f rssi_openspace=%.2f best=%.2f\n",
+			result->dist_estimates.mcpd.ifft,
+			result->dist_estimates.mcpd.phase_slope,
+			result->dist_estimates.mcpd.rssi_openspace,
+			result->dist_estimates.mcpd.best);
+	}
+}
+
+static void timeout_handler(struct k_timer *timer_id)
+{
+	sys_snode_t *node, *tmp;
+	struct peer_entry *item;
+
+	SYS_SLIST_FOR_EACH_NODE_SAFE(&peer_list, node, tmp) {
+		item = CONTAINER_OF(node, struct peer_entry, node);
+		if (item->timeout_ms > PEER_TIMEOUT_STEP_MS) {
+			item->timeout_ms -= PEER_TIMEOUT_STEP_MS;
+		} else {
+			sys_slist_remove(&peer_list, NULL, node);
+			k_heap_free(&peer_heap, item);
+			closest_peer = peer_find_closest();
+
+			led_notification(closest_peer);
+		}
+	}
+}
+
+static void peer_thread(void)
+{
+	struct dm_result result;
+
+	while (1) {
+		if (k_msgq_get(&result_msgq, &result, K_FOREVER) == 0) {
+			struct peer_entry *peer;
+
+			peer = peer_find(&result.bt_addr);
+			if (!peer) {
+				continue;
+			}
+
+			memcpy(&peer->result, &result, sizeof(peer->result));
+			peer->timeout_ms = PEER_TIMEOUT_INIT_MS;
+			print_result(&peer->result);
+
+			closest_peer = mcpd_min_peer_result(closest_peer, peer);
+
+			led_notification(closest_peer);
+			ble_notification(closest_peer);
+		}
+	}
+}
+
+bool peer_supported_test(const bt_addr_le_t *peer)
+{
+	sys_snode_t *node, *tmp;
+	struct peer_entry *item;
+
+	SYS_SLIST_FOR_EACH_NODE_SAFE(&peer_list, node, tmp) {
+		item = CONTAINER_OF(node, struct peer_entry, node);
+		if (bt_addr_le_cmp(&item->bt_addr, peer) == 0) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+int peer_supported_add(const bt_addr_le_t *peer)
+{
+	struct peer_entry *item;
+
+	if (peer_supported_test(peer)) {
+		return 0;
+	}
+
+	item = k_heap_alloc(&peer_heap, sizeof(struct peer_entry), K_NO_WAIT);
+	if (!item) {
+		return -ENOMEM;
+	}
+
+	item->timeout_ms = PEER_TIMEOUT_INIT_MS;
+	bt_addr_le_copy(&item->bt_addr, peer);
+	list_lock();
+	sys_slist_append(&peer_list, &item->node);
+	list_unlock();
+
+	return 0;
+}
+/** Prepare access address as per BT Spec.
+ *
+ * - It shall have no more than six consecutive zeros or ones.
+ * - It shall not be the advertising channel packets' Access Address.
+ * - It shall not be a sequence that differs from the advertising channel
+ *   packets Access Address by only one bit.
+ * - It shall not have all four octets equal.
+ * - It shall have no more than 24 transitions.
+ * - It shall have a minimum of two transitions in the most significant six bits.
+ */
+
+uint32_t peer_access_address_new(void)
+{
+	uint32_t access_addr;
+	uint32_t compare_access_addr;
+	uint8_t bit_idx;
+	uint8_t transitions;
+	uint8_t consecutive_cnt;
+	uint8_t consecutive_bit;
+	uint8_t same_cnt;
+
+again:
```

Comment:
Consider replacing this with while loop and continue statement for retries

Reply:
I have changed the method of preparing the access address for dm.